File: python/532.py
"""Project Euler Problem 532: Robots on a Sphere.

Suppose n robots start equidistantly on a circle of radius R on a sphere of
radius 1, and each one continuously moves toward the one after it, until they
all meet. If n is the smallest number such that each robot moves a distance
of at least N, then find the total combined distance that they all move.
"""

from __future__ import annotations

import math


def hypot(a: float, b: float) -> float:
    """Hypotenuse."""
    return math.sqrt(a * a + b * b)


def integrate(f, a: float, b: float, n: int) -> float:
    """Integrate function f from a to b using Simpson's rule."""
    h = (b - a) / n
    result = f(a) + f(b)
    for i in range(1, n):
        x = a + i * h
        if i % 2 == 0:
            result += 2 * f(x)
        else:
            result += 4 * f(x)
    return result * h / 3


def line_length(num_robots: int, R: float) -> float:
    """Compute line length for given number of robots."""
    lam = 2 * math.pi / num_robots

    def integrand(t: float) -> float:
        dlong = math.sin(lam) * math.cos(t)
        dlat = math.sin(t) * math.cos(t) * (1 - math.cos(lam))
        if dlat == 0:
            return 0.0
        return hypot(dlat, dlong) / dlat

    return integrate(integrand, math.acos(R), math.pi / 2, 10000)


def solve() -> str:
    """Solve Problem 532."""
    N = 1000
    R = 0.999

    # Binary search for smallest n such that line_length(n) >= N
    low = 1
    high = 2**30
    while low + 1 < high:
        mid = (low + high) // 2
        if line_length(mid, R) >= N:
            high = mid
        else:
            low = mid

    ans = line_length(high, R) * high
    return f"{ans:.2f}"


def main() -> None:
    """Main entry point."""
    result = solve()
    print(result)


if __name__ == "__main__":
    main()

File: c/532.c
/*
 * Project Euler 532 - Robots on a Sphere
 *
 * n robots on a circle of radius R on a unit sphere, each moving toward
 * the next. Find the smallest n such that each moves >= N=1000, then
 * compute total combined distance.
 *
 * Uses numerical integration (Simpson's rule) and binary search.
 */
#include <stdio.h>
#include <math.h>

#define STEPS 10000

static double line_length(int num_robots, double R) {
    double lam = 2.0 * M_PI / num_robots;
    double sin_lam = sin(lam);
    double cos_lam = cos(lam);
    double one_minus_cos_lam = 1.0 - cos_lam;
    double start = acos(R);
    double end = M_PI / 2.0;
    double h = (end - start) / STEPS;

    /* Simpson's rule */
    double sum = 0.0;
    for (int i = 0; i <= STEPS; i++) {
        double t = start + h * i;
        double sin_t = sin(t);
        double cos_t = cos(t);
        double dlong = sin_lam * cos_t;
        double dlat = sin_t * cos_t * one_minus_cos_lam;
        double val;
        if (fabs(dlat) < 1e-30)
            val = 0.0;
        else
            val = hypot(dlat, dlong) / dlat;

        if (i == 0 || i == STEPS)
            sum += val;
        else if (i % 2 == 1)
            sum += 4.0 * val;
        else
            sum += 2.0 * val;
    }

    return sum * h / 3.0;
}

int main(void) {
    int N_target = 1000;
    double R = 0.999;

    /* Binary search for smallest n such that line_length(n, R) >= N_target */
    int low = 1, high = 1 << 30;
    while (low + 1 < high) {
        int mid = low + (high - low) / 2;
        if (line_length(mid, R) >= N_target)
            high = mid;
        else
            low = mid;
    }

    double ans = line_length(high, R) * high;
    printf("%.2f\n", ans);
    return 0;
}
